crates/forge/src/scarb/config.rs

```diff
@@ -42,36 +42,97 @@ pub const SCARB_MANIFEST_TEMPLATE_CONTENT: &str = r#"
 
 #[allow(clippy::module_name_repetitions)]
 #[allow(clippy::struct_excessive_bools)]
-#[derive(Debug, PartialEq, Default)]
+#[derive(Debug, PartialEq, Default, Deserialize)]
 pub struct ForgeConfigFromScarb {
     /// Should runner exit after first failed test
+    #[serde(default)]
     pub exit_first: bool,
     /// How many runs should fuzzer execute
     pub fuzzer_runs: Option<NonZeroU32>,
     /// Seed to be used by fuzzer
     pub fuzzer_seed: Option<u64>,
     /// Display more detailed info about used resources
+    #[serde(default)]
     pub detailed_resources: bool,
     /// Save execution traces of all test which have passed and are not fuzz tests
+    #[serde(default)]
     pub save_trace_data: bool,
     /// Build profiles of all tests which have passed and are not fuzz tests
+    #[serde(default)]
     pub build_profile: bool,
     /// Generate a coverage report for the executed tests which have passed and are not fuzz tests
+    #[serde(default)]
     pub coverage: bool,
     /// Fork configuration profiles
+    #[serde(default, deserialize_with = "validate_forks")]
     pub fork: Vec<ForkTarget>,
     /// Limit of steps
     pub max_n_steps: Option<u32>,
 }
 
 #[non_exhaustive]
-#[derive(Debug, PartialEq, Clone)]
+#[derive(Debug, PartialEq, Clone, Deserialize)]
 pub struct ForkTarget {
     pub name: String,
     pub url: Url,
     pub block_id: BlockId,
 }
 
+fn validate_forks<'de, D>(deserializer: D) -> Result<Vec<ForkTarget>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    use serde::de::Error;
+
+    // deserialize to Vec<ForkTarget>
+    let raw_fork_targets = Vec::<RawForkTarget>::deserialize(deserializer)?;
+
+    // Validate
+    match validate_raw_fork_config(raw_fork_targets) {
+        Ok(forks) => Ok(forks),
+        Err(e) => Err(Error::custom(e.to_string())),
+    }
+}
+
+#[derive(Deserialize, Debug, PartialEq, Default, Clone)]
+pub(crate) struct RawForkTarget {
+    pub name: String,
+    pub url: String,
+    pub block_id: HashMap<String, String>,
+}
+
+fn validate_raw_fork_config(forks: Vec<RawForkTarget>) -> Result<Vec<ForkTarget>> {
+    let names: Vec<_> = forks.iter().map(|fork| &fork.name).collect();
+    let removed_duplicated_names: HashSet<_> = names.iter().collect();
+    let mut fork_targets = vec![];
+    if names.len() != removed_duplicated_names.len() {
+        bail!("Some fork names are duplicated");
+    }
+    for fork in forks {
+        let block_id_item = fork.block_id.iter().exactly_one();
+
+        let Ok((block_id_key, block_id_value)) = block_id_item else {
+            bail!("block_id should be set once per fork");
+        };
+
+        if !["number", "hash", "tag"].contains(&&**block_id_key) {
+            bail!("block_id = {block_id_key} is not valid. Possible values are = \"number\", \"hash\" and \"tag\"");
+        }
+
+        if block_id_key == "tag" && block_id_value != "latest" {
+            bail!("block_id.tag can only be equal to latest");
+        }
```

Comment:
This verification should be already handled by serde once we remove `RawForkTarget`